File: tets/test_s3.go
package main

import (
	"context"
	"fmt"
	"log"
	"os"
	"path/filepath"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/credentials"
	"github.com/aws/aws-sdk-go-v2/feature/s3/manager"
	"github.com/aws/aws-sdk-go-v2/service/s3"
)

func uploadToYandexS3(localDir, bucketName, s3Folder, accessKey, secretKey string) error {
	ctx := context.TODO()

	cfg, err := config.LoadDefaultConfig(ctx,
		config.WithCredentialsProvider(credentials.NewStaticCredentialsProvider(accessKey, secretKey, "")),
		config.WithEndpointResolver(
			aws.EndpointResolverFunc(func(service, region string) (aws.Endpoint, error) {
				return aws.Endpoint{URL: "https://storage.yandexcloud.net"}, nil
			}),
		),
		config.WithRegion("ru-central1"),
	)
	if err != nil {
		return fmt.Errorf("upload config err AWS: %v", err)
	}

	s3Client := s3.NewFromConfig(cfg)
	uploader := manager.NewUploader(s3Client)

	return filepath.Walk(localDir, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}
		if info.IsDir() {
			return nil
		}

		relPath, err := filepath.Rel(localDir, path)
		if err != nil {
			return err
		}

		s3Path := filepath.ToSlash(filepath.Join(s3Folder, relPath))
		file, err := os.Open(path)
		if err != nil {
			return err
		}
		defer file.Close()

		fmt.Printf("Uploading %s to s3://%s/%s\n", path, bucketName, s3Path)
		_, err = uploader.Upload(ctx, &s3.PutObjectInput{
			Bucket: aws.String(bucketName),
			Key:    aws.String(s3Path),
			Body:   file,
		})
		return err
	})
}

func main() {
	err := uploadToYandexS3(
		"/path/to/local/files",  
		"your-bucket-name",       
		"backup",                 
		"your-access-key",        
		"your-secret-key",        
	)
	if err != nil {
		log.Fatalf("Loading files err: %v", err)
	}
	fmt.Println("DONE")
}

File: tets/test_s3.py
import boto3
import os
from botocore.exceptions import NoCredentialsError

def upload_to_yandex_s3(local_directory: str, bucket_name: str, s3_folder: str, access_key: str, secret_key: str):
    session = boto3.session.Session()
    s3 = session.client(
        service_name='s3',
        endpoint_url='https://storage.yandexcloud.net',
        aws_access_key_id=access_key,
        aws_secret_access_key=secret_key
    )
    
    try:
        for root, _, files in os.walk(local_directory):
            for file in files:
                local_path = os.path.join(root, file)
                relative_path = os.path.relpath(local_path, local_directory)
                s3_path = f"{s3_folder}/{relative_path}".replace("\\", "/")
                
                print(f"Uploading {local_path} to s3://{bucket_name}/{s3_path}")
                s3.upload_file(local_path, bucket_name, s3_path)
                
        print("DONE")
    except NoCredentialsError:
        print("err: wrong data")
    except Exception as e:
        print(f"err: {e}")

if __name__ == "__main__":
    upload_to_yandex_s3(
        local_directory="/path/to/local/files",  
        bucket_name="your-bucket-name",        
        s3_folder="backup",                     
        access_key="your-access-key",         
        secret_key="your-secret-key"           
    )
